<!-- HTML header for doxygen 1.9.8-->
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en-US">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=11"/>
<meta name="generator" content="Doxygen 1.9.8"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>CandidateSearch: CandidateSearch</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtreedata.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="doxygen-awesome.css" rel="stylesheet" type="text/css"/>
<!-- ... other metadata & script includes ... -->
<script type="text/javascript" src="doxygen-awesome-darkmode-toggle.js"></script>
<script type="text/javascript">
    DoxygenAwesomeDarkModeToggle.init()
</script>
<script type="text/javascript" src="doxygen-awesome-fragment-copy-button.js"></script>
<script type="text/javascript">
    DoxygenAwesomeFragmentCopyButton.init()
</script>
<script type="text/javascript" src="doxygen-awesome-interactive-toc.js"></script>
<script type="text/javascript">
    DoxygenAwesomeInteractiveToc.init()
</script>
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr id="projectrow">
  <td id="projectlogo"><img alt="Logo" src="annika_logo.svg"/></td>
  <td id="projectalign">
   <div id="projectname">CandidateSearch<span id="projectnumber">&#160;1.1.2</span>
   </div>
   <div id="projectbrief">Proof-of-concept implementation of a search engine that uses sparse matrix multiplication to identify the best peptide candidates for a given mass spectrum.</div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.8 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
var searchBox = new SearchBox("searchBox", "search/",'.html');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */
</script>
<div id="main-nav"></div>
</div><!-- top -->
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
      <div id="nav-sync" class="sync"></div>
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(document).ready(function(){initNavTree('index.html',''); initResizable(); });
/* @license-end */
</script>
<div id="doc-content">
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<div id="MSearchResults">
<div class="SRPage">
<div id="SRIndex">
<div id="SRResults"></div>
<div class="SRStatus" id="Loading">Loading...</div>
<div class="SRStatus" id="Searching">Searching...</div>
<div class="SRStatus" id="NoMatches">No Matches</div>
</div>
</div>
</div>
</div>

<div><div class="header">
  <div class="headertitle"><div class="title"><a class="el" href="namespace_candidate_search.html">CandidateSearch</a> </div></div>
</div><!--header-->
<div class="contents">
<div class="textblock"><p><a class="anchor" id="md__c_1_2_users_2_p42587_2_documents_2_git_hub_2_candidate_search_2_r_e_a_d_m_e"></a> Proof-of-concept implementation of a search engine that uses <a href="https://github.com/hgb-bin-proteomics/CandidateVectorSearch">CandidateVectorSearch</a> to identify the best peptide candidates for a given mass spectrum. <em><a class="el" href="namespace_candidate_search.html">CandidateSearch</a></em> is also the computational backend of the non-cleavable crosslink search in <a href="https://github.com/hgb-bin-proteomics/MSAnnika">MS Annika</a>. <em><a class="el" href="namespace_candidate_search.html">CandidateSearch</a></em> creates the vector encodings of peptides and spectra that are needed for the sparse matrix search of <em>CandidateVectorSearch</em>.</p>
<p><em><a class="el" href="namespace_candidate_search.html">CandidateSearch</a></em> can identify peptide candidates from a given mass spectrum without any precursor ion/mass information and no previous knowledge about potential fixed or variable modifications. <em><a class="el" href="namespace_candidate_search.html">CandidateSearch</a></em> can also identify peptidoform candidates if a set of fixed and variable modifications is provided. The aim of <em><a class="el" href="namespace_candidate_search.html">CandidateSearch</a></em> is to reduce the search space for a given identification task by filtering out unlikely peptide or peptidoform candidates. It is <b>NOT</b> meant to be a standalone search engine for peptide/peptidoform identification.</p>
<p>A simplified break down of the <em><a class="el" href="namespace_candidate_search.html">CandidateSearch</a></em> algorithm is given in the following:</p><ul>
<li>Read the given MS2 spectra from the mgf file.</li>
<li>Generate the encoding vectors for each spectrum.</li>
<li>Transform spectrum encoding vectors into the representation needed for <em>CandidateVectorSearch</em>.</li>
<li>Read the given fasta file.</li>
<li>Digest the proteins of the fasta file into peptides.</li>
<li>[Optional] Generate decoy peptides and peptidoforms.</li>
<li>Calculate theoretical ion m/z values for all peptides.</li>
<li>Generate the encoding vectors for each peptide.</li>
<li>Transform the peptide encoding vectors into the representation needed for <em>CandidateVectorSearch</em>.</li>
<li>Run <em>CandidateVectorSearch</em>.</li>
<li>Process the results of <em>CandidateVectorSearch</em>.</li>
<li>Create a csv file that maps every spectrum (scan number) to a list of the best <em>n</em> peptide candidates.</li>
<li>Done!</li>
</ul>
<h1><a class="anchor" id="autotoc_md10"></a>
Usage</h1>
<p>Running <em><a class="el" href="namespace_candidate_search.html">CandidateSearch</a></em> requires three files:</p><ul>
<li>An mgf file containing MS2 spectra. <br  />
 <b>We highly recommend to deisotope and deconvolute spectra before search!</b></li>
<li>A fasta file containing sample proteins.</li>
<li>A settings file containing parameters for digestion, ion calculation and search (see below for an explanation of the settings file).</li>
</ul>
<p>The <em><a class="el" href="namespace_candidate_search.html">CandidateSearch</a></em> executable can then be run like this:</p>
<div class="fragment"><div class="line">CandidateSearch.exe spectra.mgf database.fasta settings.txt</div>
</div><!-- fragment --><p>Example files that can be used to test <em><a class="el" href="namespace_candidate_search.html">CandidateSearch</a></em> can be found in <code>/data</code>.</p>
<h1><a class="anchor" id="autotoc_md11"></a>
Settings</h1>
<p>The settings file accepts the following parameters:</p><ul>
<li>MAX_CLEAVAGES: The maximum number of allowed missed cleavages during digestion. (integer, default = 2)</li>
<li>MIN_PEP_LENGTH: The minimum length of a peptide to be considered for search. (integer, default = 5)</li>
<li>MAX_PEP_LENGTH: The maximum length of a peptide to be considered for search. (integer, default = 30)</li>
<li>MAX_PRECURSOR_CHARGE: The maximum considered precursor ion charge. (integer, default = 4)</li>
<li>MAX_FRAGMENT_CHARGE: The maximum considered fragment ion charge. (string, default = +1)</li>
<li>MAX_NEUTRAL_LOSSES: The maximum number of neutral losses considered during ion calculation. (integer, default = 1)</li>
<li>MAX_NEUTRAL_LOSS_MODS: The maximum number of neutral loss modifications considered during ion calculation. (integer, default = 2)</li>
<li>FIXED_MODIFICATIONS: Fixed modifications that should be considered during search given as <code>(char)amino_acid:(double)modification_mass</code>. An example would be carbamidomethylation of cysteine, which would be denoted as <code>C:57.021464;</code>. Several fixed modifications can be provided. (string, default = None)</li>
<li>VARIABLE_MODIFICATIONS: Variable modifications that should be considered during search given as <code>(char)amino_acid:(double)modification_mass</code>. An example would be oxidation of methionine, which would be denoted as <code>M:15.994915;</code>. Several variable modifications can be provided. If no modifications are given, <em><a class="el" href="namespace_candidate_search.html">CandidateSearch</a></em> will return the best scoring unmodified peptidoforms for a given spectrum. (string, default = None)</li>
<li>DECOY_SEARCH: Whether decoy search should be performed or not. Accepts <code>true</code> or <code>false</code>. (bool, default = true)</li>
<li>TOP_N: The number of best candidates that should be returned by the search. (integer, default = 1000)</li>
<li>TOLERANCE: Tolerance used for matching theoretical ions to experimental peaks. Given in Dalton. (double, default = 0.02)</li>
<li>NORMALIZE: Whether or not <em>CandidateVectorSearch</em> scores should be normalized before selecting the best <em>n</em> candidates. Accepts <code>true</code> or <code>false</code>. (bool, default = false)</li>
<li>USE_GAUSSIAN: Whether or not experimental peaks should be modelled as gaussian distributions with <code>mu = (m/z)</code> and <code>sigma = (tolerance/3)</code>. Accepts <code>true</code> or <code>false</code>. (bool, default = true) <br  />
</li>
<li>MODE: Search approach used by <em>CandidateVectorSearch</em>. One of the following (default = CPU_SMi32):<ul>
<li>CPU_DVi32: Sparse int matrix - dense int vector search on the CPU.</li>
<li>CPU_DVf32: Sparse float matrix - dense float vector search on the CPU.</li>
<li>CPU_DMi32: Sparse int matrix - dense int matrix search on the CPU.</li>
<li>CPU_DMf32: Sparse float matrix - dense float matrix search on the CPU.</li>
<li>CPU_SVi32: Sparse int matrix - sparse int vector search on the CPU.</li>
<li>CPU_SVf32: Sparse float matrix - sparse float vector search on the CPU.</li>
<li>CPU_SMi32: Sparse int matrix - sparse int matrix search on the CPU.</li>
<li>CPU_SMf32: Sparse float matrix - sparse float matrix search on the CPU.</li>
<li>GPU_DVf32: Sparse float matrix - dense float vector search on the GPU (see <a href="https://github.com/hgb-bin-proteomics/CandidateVectorSearch/blob/master/README.md">requirements</a>).</li>
<li>GPU_DMf32: Sparse float matrix - dense float matrix search on the GPU (see <a href="https://github.com/hgb-bin-proteomics/CandidateVectorSearch/blob/master/README.md">requirements</a>).</li>
<li>GPU_SMf32: Sparse float matrix - sparse float matrix search on the GPU (see <a href="https://github.com/hgb-bin-proteomics/CandidateVectorSearch/blob/master/README.md">requirements</a>).</li>
</ul>
</li>
</ul>
<p>For the last five parameters you might additionally want to check the documentation of <a href="https://github.com/hgb-bin-proteomics/CandidateVectorSearch">CandidateVectorSearch</a> to get a better understanding of their meaning.</p>
<p>An empty <code>settings.txt</code> file is a valid configuration for search (default parameters will be used), however not providing a settings file at all is not valid.</p>
<p>An example <code>settings.txt</code> file is provided <a href="https://github.com/hgb-bin-proteomics/CandidateSearch/blob/master/settings.txt">here</a>.</p>
<p>Additionally its contents are listed below, which should help in understanding the formatting:</p>
<div class="fragment"><div class="line">## DIGESTIONS PARAMETERS</div>
<div class="line">MAX_CLEAVAGES = 2</div>
<div class="line">MIN_PEP_LENGTH = 5</div>
<div class="line">MAX_PEP_LENGTH = 30</div>
<div class="line"> </div>
<div class="line">## ION CALCULATION PARAMETERS</div>
<div class="line">MAX_PRECURSOR_CHARGE = 4</div>
<div class="line">MAX_FRAGMENT_CHARGE = +1</div>
<div class="line">MAX_NEUTRAL_LOSSES = 1</div>
<div class="line">MAX_NEUTRAL_LOSS_MODS = 2</div>
<div class="line">#FIXED_MODIFICATIONS = None</div>
<div class="line">FIXED_MODIFICATIONS = C:57.021464;</div>
<div class="line">#VARIABLE_MODIFICATIONS = None</div>
<div class="line">VARIABLE_MODIFICATIONS = M:15.994915;</div>
<div class="line">#VARIABLE_MODIFICATIONS = M:15.994915;K:284.173607;</div>
<div class="line"> </div>
<div class="line">## SEARCH PARAMETERS</div>
<div class="line">DECOY_SEARCH = true</div>
<div class="line"> </div>
<div class="line">## VECTOR SEARCH PARAMETERS</div>
<div class="line">TOP_N = 1000</div>
<div class="line">TOLERANCE = 0.02</div>
<div class="line">NORMALIZE = false</div>
<div class="line">USE_GAUSSIAN = true</div>
<div class="line">MODE = CPU_SMi32</div>
</div><!-- fragment --><h1><a class="anchor" id="autotoc_md12"></a>
Documentation</h1>
<p>The code of this search engine is fully documented within the <code>.cs</code> code files. A good entry point is the main function of <em><a class="el" href="namespace_candidate_search.html">CandidateSearch</a></em> which is implemented <a href="https://github.com/hgb-bin-proteomics/CandidateSearch/blob/master/CandidateSearch.cs">here</a>. Documentation generated by <a href="https://github.com/doxygen/doxygen">Doxygen</a> is also available here: <a href="https://hgb-bin-proteomics.github.io/CandidateSearch/">https://hgb-bin-proteomics.github.io/CandidateSearch/</a></p>
<h1><a class="anchor" id="autotoc_md13"></a>
Requirements</h1>
<ul>
<li><a href="https://dotnet.microsoft.com/en-us/">.NET</a> may be required on Windows systems.</li>
<li>[Optional] Using GPU based approaches requires a CUDA capable GPU and CUDA version == 12.2.0 (<a href="https://developer.nvidia.com/cuda-12-2-0-download-archive">download here</a>). Other CUDA versions may or may not produce the desired results (<a href="https://github.com/hgb-bin-proteomics/CandidateVectorSearch/issues/32">see this issue</a>).</li>
</ul>
<h1><a class="anchor" id="autotoc_md14"></a>
Downloads</h1>
<p>Compiled DLLs and and executables are available in the <code>exe</code> folder or in <a href="https://github.com/hgb-bin-proteomics/CandidateSearch/releases">Releases</a>.</p>
<p>We supply compiled executables and DLLs for:</p><ul>
<li>Windows 10/11 (x86, 64-bit)</li>
<li>Ubuntu 22.04 (x86, 64-bit)</li>
<li>macOS 14.4 (arm, 64-bit)</li>
</ul>
<p>For other operating systems/architectures please compile the source code yourself! You will also need to compile <a href="https://github.com/hgb-bin-proteomics/CandidateVectorSearch">CandidateVectorSearch</a>!</p>
<h1><a class="anchor" id="autotoc_md15"></a>
Limitations</h1>
<p>This a proof-of-concept implementation that shows the applicability of our <em>CandidateVectorSearch</em> approach and not a fully fledged search engine, therefore this implementation comes with a few limitations:</p>
<ul>
<li>We currently only have implemented tryptic digestion.<ul>
<li>You can implement your own digestion <a href="https://github.com/hgb-bin-proteomics/CandidateSearch/blob/master/amanda/FASTAParser.cs#L17-L22">here</a>.</li>
</ul>
</li>
<li>We currently have not implemented support for N- or C-terminal modifications.</li>
<li>We currently have only implemented support for one possible modification per amino acid.</li>
<li>We only support spectra in centroid mode (we can't really do anything with spectra in profile mode).</li>
<li>We only support databases up to a size of 12 500 000 peptides/peptidoforms, beyond that we can't guarantee that the matrix can be allocated anymore.<ul>
<li>Consider splitting your fasta into smaller chunks and searching them separately if the generated database size exceeds 12 500 000.</li>
</ul>
</li>
<li>The limitations of <a href="https://github.com/hgb-bin-proteomics/CandidateVectorSearch?tab=readme-ov-file#limitations">CandidateVectorSearch</a> also apply here.</li>
</ul>
<h1><a class="anchor" id="autotoc_md16"></a>
Results</h1>
<p>Example results of <em><a class="el" href="namespace_candidate_search.html">CandidateSearch</a></em> and results analysis are given in <code>tests</code>. An extensive report is given in <a class="el" href="md__c_1_2_users_2_p42587_2_documents_2_git_hub_2_candidate_search_2results.html">results.md</a>.</p>
<p><img src="tests/v1.0.0/results.svg" alt="Results on a HeLa dataset" style="pointer-events: none;" class="inline"/></p>
<p><b>Figure 1:</b> Identifying peptide candidates and peptidoform candidates with <em><a class="el" href="namespace_candidate_search.html">CandidateSearch</a> [v1.0.0]</em> in a <a href="https://www.ebi.ac.uk/pride/archive/projects/PXD007750">HeLa dataset</a> using the human swissprot database. The considered ground truth was an <a href="https://ms.imp.ac.at/?goto=msamanda">MS Amanda</a> search validated with <a href="https://github.com/percolator/percolator">Percolator</a>. For every high-confidence PSM we checked if the identified peptide/peptidoform was among the top 50/100/500/1000 hits of <em><a class="el" href="namespace_candidate_search.html">CandidateSearch</a></em>. We reach almost 100% coverage within the first 1000 hits of <em><a class="el" href="namespace_candidate_search.html">CandidateSearch</a></em> (for reference: the whole database contained ~4 200 000 peptides or ~10 500 000 peptidoforms).</p>
<h1><a class="anchor" id="autotoc_md17"></a>
Benchmarks</h1>
<p>Benchmarks of the different algorithms can be found in <a class="el" href="md__c_1_2_users_2_p42587_2_documents_2_git_hub_2_candidate_search_2benchmarks.html">benchmarks.md</a>.</p>
<p><img src="benchmarks/vis_A/1A.svg" alt="benchmark_hpc_1A" style="pointer-events: none;" class="inline"/></p>
<p><b>Figure 2:</b> Int32-based sparse matrix * dense matrix search using <a href="https://eigen.tuxfamily.org/">Eigen</a> generally yields the fastest computation time on modern CPUs.</p>
<h1><a class="anchor" id="autotoc_md18"></a>
Known Issues</h1>
<p><a href="https://github.com/hgb-bin-proteomics/CandidateSearch/issues">List of known issues</a></p>
<h1><a class="anchor" id="autotoc_md19"></a>
Citing</h1>
<p>If you are using [parts of] <em><a class="el" href="namespace_candidate_search.html">CandidateSearch</a></em> please cite:</p>
<div class="fragment"><div class="line">MS Annika 3.0 (publication wip)</div>
</div><!-- fragment --><h1><a class="anchor" id="autotoc_md20"></a>
License</h1>
<ul>
<li><a href="https://github.com/hgb-bin-proteomics/CandidateSearch/blob/master/LICENSE">MIT</a></li>
</ul>
<h1><a class="anchor" id="autotoc_md21"></a>
Contact</h1>
<p><a href="#" onclick="location.href='mai'+'lto:'+'mic'+'ha'+'.bi'+'rk'+'lba'+'ue'+'r@f'+'h-'+'hag'+'en'+'ber'+'g.'+'at'; return false;">micha.birklbauer@fh-hagenberg.at</a> </p>
</div></div><!-- PageDoc -->
</div><!-- contents -->
</div><!-- doc-content -->
<!-- start footer part -->
<div id="nav-path" class="navpath"><!-- id is needed for treeview function! -->
  <ul>
    <li class="footer">Generated by <a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.9.8 </li>
  </ul>
</div>
</body>
</html>
