<!-- HTML header for doxygen 1.9.8-->
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en-US">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=11"/>
<meta name="generator" content="Doxygen 1.9.8"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>CandidateSearch: Results</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtreedata.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="doxygen-awesome.css" rel="stylesheet" type="text/css"/>
<!-- ... other metadata & script includes ... -->
<script type="text/javascript" src="doxygen-awesome-darkmode-toggle.js"></script>
<script type="text/javascript">
    DoxygenAwesomeDarkModeToggle.init()
</script>
<script type="text/javascript" src="doxygen-awesome-fragment-copy-button.js"></script>
<script type="text/javascript">
    DoxygenAwesomeFragmentCopyButton.init()
</script>
<script type="text/javascript" src="doxygen-awesome-interactive-toc.js"></script>
<script type="text/javascript">
    DoxygenAwesomeInteractiveToc.init()
</script>
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr id="projectrow">
  <td id="projectlogo"><img alt="Logo" src="annika_logo.svg"/></td>
  <td id="projectalign">
   <div id="projectname">CandidateSearch<span id="projectnumber">&#160;1.1.2</span>
   </div>
   <div id="projectbrief">Proof-of-concept implementation of a search engine that uses sparse matrix multiplication to identify the best peptide candidates for a given mass spectrum.</div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.8 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
var searchBox = new SearchBox("searchBox", "search/",'.html');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */
</script>
<div id="main-nav"></div>
</div><!-- top -->
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
      <div id="nav-sync" class="sync"></div>
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(document).ready(function(){initNavTree('md__c_1_2_users_2_p42587_2_documents_2_git_hub_2_candidate_search_2results.html',''); initResizable(); });
/* @license-end */
</script>
<div id="doc-content">
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<div id="MSearchResults">
<div class="SRPage">
<div id="SRIndex">
<div id="SRResults"></div>
<div class="SRStatus" id="Loading">Loading...</div>
<div class="SRStatus" id="Searching">Searching...</div>
<div class="SRStatus" id="NoMatches">No Matches</div>
</div>
</div>
</div>
</div>

<div><div class="header">
  <div class="headertitle"><div class="title">Results</div></div>
</div><!--header-->
<div class="contents">
<div class="textblock"><p><a class="anchor" id="autotoc_md22"></a> In order to assess the applicability of our candidate search, we first tested the algorithm on linear peptides. This showed very good results, especially with deconvoluted data. Moreover, we then also applied the algorithm to non-cleavable crosslink data and once more saw good results.</p>
<h1><a class="anchor" id="autotoc_md23"></a>
Test Methodology</h1>
<p>For testing against linear peptides, mass spectrometry RAW data of HeLa cells was retrieved from PRIDE via identifier <a href="https://www.ebi.ac.uk/pride/archive/projects/PXD007750">PXD007750</a> and then exported to mgf format with Proteome Discoverer 3.1, either directly or with deisotoping and charge deconvolution. For comparison we searched the RAW data with <a href="https://ms.imp.ac.at/?goto=msamanda">MS Amanda</a> (version 3.1.21.45, Engine version 3.0.21.45, see search settings in Table 1) and validated the results with <a href="https://github.com/percolator/percolator">Percolator</a> (version 3.05.0) for 1% estimated false discovery rate (FDR). For every high-confidence peptide spectrum match (PSM) we then checked if the associated peptide was within the <em>top N</em> peptide candidates returned by the algorithm.</p>
<p>The used database was <code>uniprotkb_proteome_UP000005640_AND_revi_2024_03_22.fasta</code> (Human SwissProt).</p>
<table class="markdownTable">
<tr class="markdownTableHead">
<th class="markdownTableHeadLeft">Parameter   </th><th class="markdownTableHeadLeft">Value    </th></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyLeft">MS1 Tolerance   </td><td class="markdownTableBodyLeft">5 ppm    </td></tr>
<tr class="markdownTableRowEven">
<td class="markdownTableBodyLeft">MS2 Tolerance   </td><td class="markdownTableBodyLeft">10 ppm    </td></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyLeft">Max. Missed Cleavages   </td><td class="markdownTableBodyLeft">2    </td></tr>
<tr class="markdownTableRowEven">
<td class="markdownTableBodyLeft">Minimum Peptide Length   </td><td class="markdownTableBodyLeft">5    </td></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyLeft">Maximum Peptide Length   </td><td class="markdownTableBodyLeft">30    </td></tr>
<tr class="markdownTableRowEven">
<td class="markdownTableBodyLeft">Fixed Modification   </td><td class="markdownTableBodyLeft">Carbamidomethylation(C)    </td></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyLeft">Variable Modification   </td><td class="markdownTableBodyLeft">Oxidation(M)   </td></tr>
</table>
<p><b>Table 1:</b> Search settings used for <a href="https://ms.imp.ac.at/?goto=msamanda">MS Amanda</a> to identify PSMs.</p>
<p>For testing against cross-linked peptides, mass spectrometry RAW data was retrieved from PRIDE via identifier <a href="https://www.ebi.ac.uk/pride/archive/projects/PXD014337">PXD014337</a> and exported the same way. For comparison we used available results from the cross-linking search engine <a href="https://doi.org/10.1021/acs.analchem.1c03688">MaxLynx</a> which were also retrieved from PRIDE via identifier <a href="https://www.ebi.ac.uk/pride/archive/projects/PXD027159">PXD027159</a>. Analogously, we checked for every high-confidence (1% FDR) crosslink spectrum match (CSM) if one of the associated peptides was within the <em>top N</em> peptide candidates returned by the algorithm.</p>
<p>The used database was <code>cas9_uniprotkb_proteome_UP000005640_AND_revi_2024_03_22.fasta</code> (Human SwissProt + S. pyogenes Cas9).</p>
<h1><a class="anchor" id="autotoc_md24"></a>
[r0a] Normalize = off, Gaussian = on</h1>
<p>Before analysing the complete datasets we studied the influence of the parameters <code>NORMALIZE</code> and <code>USE_GAUSSIAN</code>. The following plots show the results using <code>NORMALIZE = false</code> and <code>USE_GAUSSIAN = true</code> for replicate 1 using eiter RAW or deconvoluted spectra.</p>
<h2><a class="anchor" id="autotoc_md25"></a>
[r0a1] raw</h2>
<p><img src="tests/v1.1.2/r0a1/r0a1.svg" alt="r0a1" style="pointer-events: none;" class="inline"/></p>
<p><b>Figure 1:</b> Results for <a href="https://www.ebi.ac.uk/pride/archive/projects/PXD007750">PXD007750</a> replicate 1 (RAW) using <code>NORMALIZE = false</code> and <code>USE_GAUSSIAN = true</code>.</p>
<h2><a class="anchor" id="autotoc_md26"></a>
[r0a2] deconvoluted</h2>
<p><img src="tests/v1.1.2/r0a2/r0a2.svg" alt="r0a2" style="pointer-events: none;" class="inline"/></p>
<p><b>Figure 2:</b> Results for <a href="https://www.ebi.ac.uk/pride/archive/projects/PXD007750">PXD007750</a> replicate 1 (deconvoluted) using <code>NORMALIZE = false</code> and <code>USE_GAUSSIAN = true</code>.</p>
<h1><a class="anchor" id="autotoc_md27"></a>
[r0b] Normalize = off, Gaussian = off</h1>
<h2><a class="anchor" id="autotoc_md28"></a>
[r0b1] raw</h2>
<p><img src="tests/v1.1.2/r0b1/r0b1.svg" alt="r0b1" style="pointer-events: none;" class="inline"/></p>
<p><b>Figure 3:</b> Results for <a href="https://www.ebi.ac.uk/pride/archive/projects/PXD007750">PXD007750</a> replicate 1 (RAW) using <code>NORMALIZE = false</code> and <code>USE_GAUSSIAN = false</code>.</p>
<h2><a class="anchor" id="autotoc_md29"></a>
[r0b2] deconvoluted</h2>
<p><img src="tests/v1.1.2/r0b2/r0b2.svg" alt="r0b2" style="pointer-events: none;" class="inline"/></p>
<p><b>Figure 4:</b> Results for <a href="https://www.ebi.ac.uk/pride/archive/projects/PXD007750">PXD007750</a> replicate 1 (deconvoluted) using <code>NORMALIZE = false</code> and <code>USE_GAUSSIAN = false</code>.</p>
<h1><a class="anchor" id="autotoc_md30"></a>
[r0c] Normalize = on, Gaussian = off</h1>
<h2><a class="anchor" id="autotoc_md31"></a>
[r0c1] raw</h2>
<p><img src="tests/v1.1.2/r0c1/r0c1.svg" alt="r0c1" style="pointer-events: none;" class="inline"/></p>
<p><b>Figure 5:</b> Results for <a href="https://www.ebi.ac.uk/pride/archive/projects/PXD007750">PXD007750</a> replicate 1 (RAW) using <code>NORMALIZE = true</code> and <code>USE_GAUSSIAN = false</code>.</p>
<h2><a class="anchor" id="autotoc_md32"></a>
[r0c2] deconvoluted</h2>
<p><img src="tests/v1.1.2/r0c2/r0c2.svg" alt="r0c2" style="pointer-events: none;" class="inline"/></p>
<p><b>Figure 6:</b> Results for <a href="https://www.ebi.ac.uk/pride/archive/projects/PXD007750">PXD007750</a> replicate 1 (deconvoluted) using <code>NORMALIZE = true</code> and <code>USE_GAUSSIAN = false</code>.</p>
<h1><a class="anchor" id="autotoc_md33"></a>
[r0d] Normalize = on, Gaussian = on</h1>
<h2><a class="anchor" id="autotoc_md34"></a>
[r0d1] raw</h2>
<p><img src="tests/v1.1.2/r0d1/r0d1.svg" alt="r0d1" style="pointer-events: none;" class="inline"/></p>
<p><b>Figure 7:</b> Results for <a href="https://www.ebi.ac.uk/pride/archive/projects/PXD007750">PXD007750</a> replicate 1 (RAW) using <code>NORMALIZE = true</code> and <code>USE_GAUSSIAN = true</code>.</p>
<h2><a class="anchor" id="autotoc_md35"></a>
[r0d2] deconvoluted</h2>
<p><img src="tests/v1.1.2/r0d2/r0d2.svg" alt="r0d2" style="pointer-events: none;" class="inline"/></p>
<p><b>Figure 8:</b> Results for <a href="https://www.ebi.ac.uk/pride/archive/projects/PXD007750">PXD007750</a> replicate 1 (deconvoluted) using <code>NORMALIZE = true</code> and <code>USE_GAUSSIAN = true</code>.</p>
<h1><a class="anchor" id="autotoc_md36"></a>
[r1] HeLa</h1>
<p>It is pretty clear from <em>r0a1</em> to <em>r0d2</em> that parameter combination <code>NORMALIZE = true</code> and <code>USE_GAUSSIAN = true</code> with deconvoluted spectra yields the best results. This is what we therefore used for final analysis of all three replicates of the dataset.</p>
<h2><a class="anchor" id="autotoc_md37"></a>
[r1a] rep 1</h2>
<p><img src="tests/v1.1.2/r1a/r1a.svg" alt="r1a" style="pointer-events: none;" class="inline"/></p>
<p><b>Figure 9:</b> Results for <a href="https://www.ebi.ac.uk/pride/archive/projects/PXD007750">PXD007750</a> replicate 1 (deconvoluted) using <code>NORMALIZE = false</code> and <code>USE_GAUSSIAN = true</code>.</p>
<h2><a class="anchor" id="autotoc_md38"></a>
[r1b] rep 2</h2>
<p><img src="tests/v1.1.2/r1b/r1b.svg" alt="r1b" style="pointer-events: none;" class="inline"/></p>
<p><b>Figure 10:</b> Results for <a href="https://www.ebi.ac.uk/pride/archive/projects/PXD007750">PXD007750</a> replicate 2 (deconvoluted) using <code>NORMALIZE = false</code> and <code>USE_GAUSSIAN = true</code>.</p>
<h2><a class="anchor" id="autotoc_md39"></a>
[r1c] rep 3</h2>
<p><img src="tests/v1.1.2/r1c/r1c.svg" alt="r1c" style="pointer-events: none;" class="inline"/></p>
<p><b>Figure 11:</b> Results for <a href="https://www.ebi.ac.uk/pride/archive/projects/PXD007750">PXD007750</a> replicate 3 (deconvoluted) using <code>NORMALIZE = false</code> and <code>USE_GAUSSIAN = true</code>.</p>
<h1><a class="anchor" id="autotoc_md40"></a>
[r2] Beveridge</h1>
<p>For the cross-linking data we used the same settings as for linear peptides: <code>NORMALIZE = false</code> and <code>USE_GAUSSIAN = true</code> using deconvoluted spectra.</p>
<h2><a class="anchor" id="autotoc_md41"></a>
[r2a] rep 1</h2>
<p><img src="tests/v1.1.2/r2a/r2a.svg" alt="r2a" style="pointer-events: none;" class="inline"/></p>
<p><b>Figure 12:</b> Results for <a href="https://www.ebi.ac.uk/pride/archive/projects/PXD014337">PXD014337</a> replicate 1 (deconvoluted) using <code>NORMALIZE = false</code> and <code>USE_GAUSSIAN = true</code>.</p>
<h2><a class="anchor" id="autotoc_md42"></a>
[r2b] rep 2</h2>
<p><img src="tests/v1.1.2/r2b/r2b.svg" alt="r2b" style="pointer-events: none;" class="inline"/></p>
<p><b>Figure 13:</b> Results for <a href="https://www.ebi.ac.uk/pride/archive/projects/PXD014337">PXD014337</a> replicate 2 (deconvoluted) using <code>NORMALIZE = false</code> and <code>USE_GAUSSIAN = true</code>.</p>
<h2><a class="anchor" id="autotoc_md43"></a>
[r2c] rep 3</h2>
<p><img src="tests/v1.1.2/r2c/r2c.svg" alt="r2c" style="pointer-events: none;" class="inline"/></p>
<p><b>Figure 14:</b> Results for <a href="https://www.ebi.ac.uk/pride/archive/projects/PXD014337">PXD014337</a> replicate 3 (deconvoluted) using <code>NORMALIZE = false</code> and <code>USE_GAUSSIAN = true</code>.</p>
<h1><a class="anchor" id="autotoc_md44"></a>
Data Availability</h1>
<p>The full list of files for these tests can be accessed via <a href="http://u.pc.cd/z75otalK">http://u.pc.cd/z75otalK</a>.</p>
<h1><a class="anchor" id="autotoc_md45"></a>
Conclusion</h1>
<p>We could show that both for linear peptides and cross-linked peptides our algorithm is capable of finding the correct peptide candidate for identification. Interestingly, normalization does not improve results, quite contrary they get a lot worse. The best results were achieved using deconvoluted spectra with parameter settings <code>NORMALIZE = false</code> and <code>USE_GAUSSIAN = true</code>. </p>
</div></div><!-- contents -->
</div><!-- PageDoc -->
</div><!-- doc-content -->
<!-- start footer part -->
<div id="nav-path" class="navpath"><!-- id is needed for treeview function! -->
  <ul>
    <li class="footer">Generated by <a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.9.8 </li>
  </ul>
</div>
</body>
</html>
