\chapter{Candidate\+Search}
\hypertarget{md__r_e_a_d_m_e}{}\label{md__r_e_a_d_m_e}\index{CandidateSearch@{CandidateSearch}}
\label{md__r_e_a_d_m_e_autotoc_md0}%
\Hypertarget{md__r_e_a_d_m_e_autotoc_md0}%
 Proof-\/of-\/concept implementation of a search engine that uses \href{https://github.com/hgb-bin-proteomics/CandidateVectorSearch}{\texttt{ Candidate\+Vector\+Search}} to identify the best peptide candidates for a given mass spectrum. {\itshape \doxylink{namespace_candidate_search}{Candidate\+Search}} creates the vector encodings of peptides and spectra that are needed for the sparse matrix search of {\itshape Candidate\+Vector\+Search}.

{\itshape \doxylink{namespace_candidate_search}{Candidate\+Search}} can identify peptide candidates from a given mass spectrum without any precursor ion/mass information and no previous knowledge about potential fixed or variable modifications. {\itshape \doxylink{namespace_candidate_search}{Candidate\+Search}} can also identify peptidoform candidates if a set of fixed and variable modifications is provided. The aim of {\itshape \doxylink{namespace_candidate_search}{Candidate\+Search}} is to reduce the search space for a given identification task by filtering out unlikely peptide or peptidoform candidates. It is {\bfseries{NOT}} meant to be a standalone search engine for peptide/peptidoform identification.

A simplified break down of the {\itshape \doxylink{namespace_candidate_search}{Candidate\+Search}} algorithm is given in the following\+:
\begin{DoxyItemize}
\item Read the given MS2 spectra from the mgf file.
\item Generate the encoding vectors for each spectrum.
\item Transform spectrum encoding vectors into the representation needed for {\itshape Candidate\+Vector\+Search}.
\item Read the given fasta file.
\item Digest the proteins of the fasta file into peptides.
\item \mbox{[}Optional\mbox{]} Generate decoy peptides and peptidoforms.
\item Calculate theoretical ion m/z values for all peptides.
\item Generate the encoding vectors for each peptide.
\item Transform the peptide encoding vectors into the representation needed for {\itshape Candidate\+Vector\+Search}.
\item Run {\itshape Candidate\+Vector\+Search}.
\item Process the results of {\itshape Candidate\+Vector\+Search}.
\item Create a csv file that maps every spectrum (scan number) to a list of the best {\itshape n} peptide candidates.
\item Done!
\end{DoxyItemize}\hypertarget{md__r_e_a_d_m_e_autotoc_md1}{}\doxysection{\texorpdfstring{Usage}{Usage}}\label{md__r_e_a_d_m_e_autotoc_md1}
Running {\itshape \doxylink{namespace_candidate_search}{Candidate\+Search}} requires three files\+:
\begin{DoxyItemize}
\item An mgf file containing MS2 spectra.
\item A fasta file containing sample proteins.
\item A settings file containing parameters for digestion, ion calculation and search (see below for an explanation of the settings file).
\end{DoxyItemize}

The {\itshape \doxylink{namespace_candidate_search}{Candidate\+Search}} executable can then be run like this\+:


\begin{DoxyCode}{0}
\DoxyCodeLine{CandidateSearch.exe\ spectra.mgf\ database.fasta\ settings.txt}

\end{DoxyCode}


Example files that can be used to test {\itshape \doxylink{namespace_candidate_search}{Candidate\+Search}} can be found in {\ttfamily /data}.\hypertarget{md__r_e_a_d_m_e_autotoc_md2}{}\doxysection{\texorpdfstring{Settings}{Settings}}\label{md__r_e_a_d_m_e_autotoc_md2}
The settings file accepts the following parameters\+:
\begin{DoxyItemize}
\item MAX\+\_\+\+CLEAVAGES\+: The maximum number of allowed missed cleavages during digestion. (integer, default = 2)
\item MIN\+\_\+\+PEP\+\_\+\+LENGTH\+: The minimum length of a peptide to be considered for search. (integer, default = 5)
\item MAX\+\_\+\+PEP\+\_\+\+LENGTH\+: The maximum length of a peptide to be considered for search. (integer, default = 30)
\item MAX\+\_\+\+PRECURSOR\+\_\+\+CHARGE\+: The maximum considered precursor ion charge. (integer, default = 4)
\item MAX\+\_\+\+FRAGMENT\+\_\+\+CHARGE\+: The maximum considered fragment ion charge. (string, default = +1)
\item MAX\+\_\+\+NEUTRAL\+\_\+\+LOSSES\+: The maximum number of neutral losses considered during ion calculation. (integer, default = 1)
\item MAX\+\_\+\+NEUTRAL\+\_\+\+LOSS\+\_\+\+MODS\+: The maximum number of neutral loss modifications considered during ion calculation. (integer, default = 2)
\item FIXED\+\_\+\+MODIFICATIONS\+: Fixed modifications that should be considered during search given as {\ttfamily (char)amino\+\_\+acid\+:(double)modification\+\_\+mass}. An example would be carbamidomethylation of cysteine, which would be denoted as {\ttfamily C\+:57.\+021464;}. Several fixed modifications can be provided. (string, default = None)
\item VARIABLE\+\_\+\+MODIFICATIONS\+: Variable modifications that should be considered during search given as {\ttfamily (char)amino\+\_\+acid\+:(double)modification\+\_\+mass}. An example would be oxidation of methionine, which would be denoted as {\ttfamily M\+:15.\+994915;}. Several variable modifications can be provided. If no modifications are given, {\itshape \doxylink{namespace_candidate_search}{Candidate\+Search}} will return the best scoring unmodified peptidoforms for a given spectrum. (string, default = None)
\item DECOY\+\_\+\+SEARCH\+: Whether decoy search should be performed or not. Accepts {\ttfamily true} or {\ttfamily false}. (bool, default = true)
\item TOP\+\_\+N\+: The number of best candidates that should be returned by the search. (integer, default = 1000)
\item TOLERANCE\+: Tolerance used for matching theoretical ions to experimental peaks. Given in Dalton. (double, default = 0.\+02)
\item NORMALIZE\+: Whether or not {\itshape Candidate\+Vector\+Search} scores should be normalized before selecting the best {\itshape n} candidates. Accepts {\ttfamily true} or {\ttfamily false}. (bool, default = false)
\item USE\+\_\+\+GAUSSIAN\+: Whether or not experimental peaks should be modelled as gaussian distributions with {\ttfamily mu = (m/z)} and {\ttfamily sigma = (tolerance/3)}. Accepts {\ttfamily true} or {\ttfamily false}. (bool, default = true) ~\newline

\item MODE\+: Search approach used by {\itshape Candidate\+Vector\+Search}. One of the following (default = CPU\+\_\+\+DV)\+:
\begin{DoxyItemize}
\item CPU\+\_\+\+DV\+: Sparse matrix -\/ dense vector search on the CPU.
\item CPU\+\_\+\+DM\+: Sparse matrix -\/ dense matrix search on the CPU.
\item CPU\+\_\+\+SV\+: Sparse matrix -\/ sparse vector search on the CPU.
\item CPU\+\_\+\+SM\+: Sparse matrix -\/ sparse matrix search on the CPU.
\item GPU\+\_\+\+DV\+: Sparse matrix -\/ dense vector search on the GPU (see \href{https://github.com/hgb-bin-proteomics/CandidateVectorSearch/blob/master/README.md}{\texttt{ requirements}}).
\item GPU\+\_\+\+DM\+: Sparse matrix -\/ dense matrix search on the GPU (see \href{https://github.com/hgb-bin-proteomics/CandidateVectorSearch/blob/master/README.md}{\texttt{ requirements}}).
\item GPU\+\_\+\+SM\+: Sparse matrix -\/ sparse matrix search on the GPU (see \href{https://github.com/hgb-bin-proteomics/CandidateVectorSearch/blob/master/README.md}{\texttt{ requirements}}).
\end{DoxyItemize}
\end{DoxyItemize}

For the last five parameters you might additionally want to check the documentation of \href{https://github.com/hgb-bin-proteomics/CandidateVectorSearch}{\texttt{ Candidate\+Vector\+Search}} to get a better understanding of their meaning.

An empty {\ttfamily settings.\+txt} file is a valid configuration for search (default parameters will be used), however not providing a settings file at all is not valid.

An example {\ttfamily settings.\+txt} file is provided \href{https://github.com/hgb-bin-proteomics/CandidateSearch/blob/master/settings.txt}{\texttt{ here}}.

Additionally its contents are listed below, which should help in understanding the formatting\+:


\begin{DoxyCode}{0}
\DoxyCodeLine{\#\#\ DIGESTIONS\ PARAMETERS}
\DoxyCodeLine{MAX\_CLEAVAGES\ =\ 2}
\DoxyCodeLine{MIN\_PEP\_LENGTH\ =\ 5}
\DoxyCodeLine{MAX\_PEP\_LENGTH\ =\ 30}
\DoxyCodeLine{}
\DoxyCodeLine{\#\#\ ION\ CALCULATION\ PARAMETERS}
\DoxyCodeLine{MAX\_PRECURSOR\_CHARGE\ =\ 4}
\DoxyCodeLine{MAX\_FRAGMENT\_CHARGE\ =\ +1}
\DoxyCodeLine{MAX\_NEUTRAL\_LOSSES\ =\ 1}
\DoxyCodeLine{MAX\_NEUTRAL\_LOSS\_MODS\ =\ 2}
\DoxyCodeLine{\#FIXED\_MODIFICATIONS\ =\ None}
\DoxyCodeLine{FIXED\_MODIFICATIONS\ =\ C:57.021464;}
\DoxyCodeLine{\#VARIABLE\_MODIFICATIONS\ =\ None}
\DoxyCodeLine{VARIABLE\_MODIFICATIONS\ =\ M:15.994915;}
\DoxyCodeLine{\#VARIABLE\_MODIFICATIONS\ =\ M:15.994915;K:284.173607;}
\DoxyCodeLine{}
\DoxyCodeLine{\#\#\ SEARCH\ PARAMETERS}
\DoxyCodeLine{DECOY\_SEARCH\ =\ true}
\DoxyCodeLine{}
\DoxyCodeLine{\#\#\ VECTOR\ SEARCH\ PARAMETERS}
\DoxyCodeLine{TOP\_N\ =\ 1000}
\DoxyCodeLine{TOLERANCE\ =\ 0.02}
\DoxyCodeLine{NORMALIZE\ =\ false}
\DoxyCodeLine{USE\_GAUSSIAN\ =\ true}
\DoxyCodeLine{MODE\ =\ CPU\_DV}

\end{DoxyCode}
\hypertarget{md__r_e_a_d_m_e_autotoc_md3}{}\doxysection{\texorpdfstring{Documentation}{Documentation}}\label{md__r_e_a_d_m_e_autotoc_md3}
The code of this search engine is fully documented within the {\ttfamily .cs} code files. A good entry point is the main function of {\itshape \doxylink{namespace_candidate_search}{Candidate\+Search}} which is implemented \href{https://github.com/hgb-bin-proteomics/CandidateSearch/blob/master/CandidateSearch.cs}{\texttt{ here}}. Documentation generated by \href{https://github.com/doxygen/doxygen}{\texttt{ Doxygen}} is also available here\+: \href{https://hgb-bin-proteomics.github.io/CandidateSearch/}{\texttt{ https\+://hgb-\/bin-\/proteomics.\+github.\+io/\+Candidate\+Search/}}\hypertarget{md__r_e_a_d_m_e_autotoc_md4}{}\doxysection{\texorpdfstring{Limitations}{Limitations}}\label{md__r_e_a_d_m_e_autotoc_md4}
This a proof-\/of-\/concept implementation that shows the applicability of our {\itshape Candidate\+Vector\+Search} approach and not a fully fledged search engine, therefore this implementation comes with a few limitations\+:


\begin{DoxyItemize}
\item We currently only have implemented tryptic digestion.
\begin{DoxyItemize}
\item You can implement your own digestion \href{https://github.com/hgb-bin-proteomics/CandidateSearch/blob/master/amanda/FASTAParser.cs\#L17-L22}{\texttt{ here}}.
\end{DoxyItemize}
\item We currently have not implemented support for N-\/ or C-\/terminal modifications.
\item We currently have only implemented support for one possible modification per amino acid.
\item We only support spectra in centroid mode (we can\textquotesingle{}t really do anything with spectra in profile mode).
\end{DoxyItemize}\hypertarget{md__r_e_a_d_m_e_autotoc_md5}{}\doxysection{\texorpdfstring{Results}{Results}}\label{md__r_e_a_d_m_e_autotoc_md5}
Example results of {\itshape \doxylink{namespace_candidate_search}{Candidate\+Search}} and results analysis are given in {\ttfamily tests}.



{\bfseries{Figure 1\+:}} Identifying peptide candidates and peptidoform candidates with {\itshape \doxylink{namespace_candidate_search}{Candidate\+Search}} in a \href{https://www.ebi.ac.uk/pride/archive/projects/PXD007750}{\texttt{ He\+La dataset}} using the human swissprot database. The considered ground truth was an \href{https://ms.imp.ac.at/?goto=msamanda}{\texttt{ MS Amanda}} search validated with \href{https://github.com/percolator/percolator}{\texttt{ Percolator}}. For every high-\/confidence PSM we checked if the identified peptide/peptidoform was among the top 50/100/500/1000 hits of {\itshape \doxylink{namespace_candidate_search}{Candidate\+Search}}. We reach almost 100\% coverage within the first 1000 hits of {\itshape \doxylink{namespace_candidate_search}{Candidate\+Search}} (for reference\+: the whole database contained \texorpdfstring{$\sim$}{\string~}4 200 000 peptides or \texorpdfstring{$\sim$}{\string~}10 500 000 peptidoforms).\hypertarget{md__r_e_a_d_m_e_autotoc_md6}{}\doxysection{\texorpdfstring{Known Issues}{Known Issues}}\label{md__r_e_a_d_m_e_autotoc_md6}
\href{https://github.com/hgb-bin-proteomics/CandidateSearch/issues}{\texttt{ List of known issues}}\hypertarget{md__r_e_a_d_m_e_autotoc_md7}{}\doxysection{\texorpdfstring{Citing}{Citing}}\label{md__r_e_a_d_m_e_autotoc_md7}
If you are using \mbox{[}parts of\mbox{]} {\itshape \doxylink{namespace_candidate_search}{Candidate\+Search}} please cite\+:


\begin{DoxyCode}{0}
\DoxyCodeLine{MS\ Annika\ 3.0\ (publication\ wip)}

\end{DoxyCode}
\hypertarget{md__r_e_a_d_m_e_autotoc_md8}{}\doxysection{\texorpdfstring{License}{License}}\label{md__r_e_a_d_m_e_autotoc_md8}

\begin{DoxyItemize}
\item \href{https://github.com/hgb-bin-proteomics/CandidateSearch/blob/master/LICENSE}{\texttt{ MIT}}
\end{DoxyItemize}\hypertarget{md__r_e_a_d_m_e_autotoc_md9}{}\doxysection{\texorpdfstring{Contact}{Contact}}\label{md__r_e_a_d_m_e_autotoc_md9}
\href{mailto:micha.birklbauer@fh-hagenberg.at}{\texttt{ micha.\+birklbauer@fh-\/hagenberg.\+at}} 